File: ios/OmnichannelVoiceVideoCallingReactNative.m
#import <React/RCTBridgeModule.h>
#import <React/RCTViewManager.h>
#import "React/RCTEventEmitter.h"

@interface RCT_EXTERN_MODULE(OmnichannelVoiceVideoCallingReactNative, NSObject)

RCT_EXTERN_METHOD(initialize:(NSString)callingToken
                  withRequestId:(NSString)requestId
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(acceptCall:(BOOL)withVideo
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(isMicrophoneMuted:
                  (RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(toggleMute:
                  (RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(isRemoteVideoEnabled:
                  (RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(isLocalVideoEnabled:
                  (RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)


RCT_EXTERN_METHOD(rejectCall)

RCT_EXTERN_METHOD(stopCall)

RCT_EXTERN_METHOD(toggleSpeaker)

RCT_EXTERN_METHOD(toggleLocalVideo)

RCT_EXTERN_METHOD(toggleCamera)

@end

@interface RCT_EXTERN_MODULE(LocalVideoViewManager, RCTViewManager)
@end

@interface RCT_EXTERN_MODULE(RemoteVideoViewManager, RCTViewManager)
@end

File: ios/OmnichannelVoiceVideoCallingReactNative.swift
@objc(OmnichannelVoiceVideoCallingReactNative)
class OmnichannelVoiceVideoCallingReactNative: NSObject {

    private static var omnichannelVoiceVideoInstance : OmnichannelVoiceVideoCallingReactNative!

    static func getInstance() -> OmnichannelVoiceVideoCallingReactNative {
        if (self.omnichannelVoiceVideoInstance == nil){
            self.omnichannelVoiceVideoInstance = OmnichannelVoiceVideoCallingReactNative()
        }

      return self.omnichannelVoiceVideoInstance
    }

    @objc(initialize:withRequestId:withResolver:withRejecter:)
    func initialize(callingToken: String, requestId: String, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        VoiceVideoCalling.getInstance().initialize(callingToken:callingToken, requestId:requestId)
    }

    @objc(acceptCall:withResolver:withRejecter:)
    func acceptCall(withVideo: Bool, resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void
    {
        VoiceVideoCalling.getInstance().acceptCall(withVideo: withVideo)
    }

    @objc(toggleMute:rejecter:)
    func toggleMute(_ resolve: RCTPromiseResolveBlock, rejecter reject:RCTPromiseRejectBlock) -> Void {
        resolve(VoiceVideoCalling.getInstance().toggleMute())
    }

    @objc public func rejectCall() -> Void
    {
        VoiceVideoCalling.getInstance().rejectCall()
    }

    @objc public func stopCall() -> Void
    {
        VoiceVideoCalling.getInstance().stopCall()
    }

    @objc public func toggleSpeaker() -> Void
    {
        VoiceVideoCalling.getInstance().toggleSpeaker()
    }

    @objc public func toggleLocalVideo() -> Void
    {
        VoiceVideoCalling.getInstance().toggleLocalVideo()
    }

    @objc public func toggleCamera() -> Void
    {
        VoiceVideoCalling.getInstance().toggleCamera()
    }

    @objc(isMicrophoneMuted:rejecter:)
    func isMicrophoneMuted(_ resolve: RCTPromiseResolveBlock, rejecter reject:RCTPromiseRejectBlock) -> Void {
        resolve(VoiceVideoCalling.getInstance().isMicrophoneMuted())
    }

    @objc(isRemoteVideoEnabled:rejecter:)
    func isRemoteVideoEnabled(_ resolve: RCTPromiseResolveBlock, rejecter reject:RCTPromiseRejectBlock) -> Void {
        resolve(VoiceVideoCalling.getInstance().isRemoteVideoEnabled())
    }

    @objc(isLocalVideoEnabled:rejecter:)
    func isLocalVideoEnabled(_ resolve: RCTPromiseResolveBlock, rejecter reject:RCTPromiseRejectBlock) -> Void {
        resolve(VoiceVideoCalling.getInstance().isLocalVideoEnabled())
    }
}
